File: src/store/gameStore.ts
import { create } from 'zustand'
import { persist } from 'zustand/middleware'
import { applyScan, createHunt, markReady, setStageHint, startHunt } from '../lib/hunt'
import { awardTreasure } from '../lib/treasures'
import type { CollectedTreasure, Hunt, ScanResult, Screen } from '../types/game'

const INITIAL_SCREEN: Screen = 'home'

interface GameState {
  screen: Screen
  hunt: Hunt | null
  setupIndex: number
  collection: CollectedTreasure[]
  lastAwardedId: string | null
  scanResult: ScanResult | null
  practiceMode: boolean
  goTo: (screen: Screen) => void
  beginSetup: () => void
  chooseStageCount: (count: number) => void
  updateCurrentHint: (hint: string) => void
  nextSetupStage: () => void
  prevSetupStage: () => void
  finishSetup: () => void
  startAdventure: () => void
  openScan: () => void
  closeScan: () => void
  scanPayload: (payload: string) => ScanResult | null
  clearScanResult: () => void
  openCollection: () => void
  finishClear: () => void
  abandonHunt: () => void
  setPracticeMode: (value: boolean) => void
  reviewQr: () => void
}

export const useGameStore = create<GameState>()(
  persist(
    (set, get) => ({
      screen: INITIAL_SCREEN,
      hunt: null,
      setupIndex: 1,
      collection: [],
      lastAwardedId: null,
      scanResult: null,
      practiceMode: false,
      goTo: (screen) => set({ screen, scanResult: null }),
      beginSetup: () =>
        set({
          screen: 'setup-count',
          hunt: null,
          setupIndex: 1,
          scanResult: null,
        }),
      chooseStageCount: (count) =>
        set({
          hunt: createHunt(count),
          setupIndex: 1,
          screen: 'setup-stage',
        }),
      updateCurrentHint: (hint) => {
        const { hunt, setupIndex } = get()
        if (!hunt) {
          return
        }
        set({ hunt: setStageHint(hunt, setupIndex, hint) })
      },
      nextSetupStage: () => {
        const { hunt, setupIndex } = get()
        if (!hunt) {
          return
        }
        if (setupIndex >= hunt.stageCount) {
          get().finishSetup()
          return
        }
        set({ setupIndex: setupIndex + 1 })
      },
      prevSetupStage: () => {
        const { setupIndex } = get()
        if (setupIndex <= 1) {
          set({ screen: 'setup-count' })
          return
        }
        set({ setupIndex: setupIndex - 1 })
      },
      finishSetup: () => {
        const { hunt } = get()
        if (!hunt) {
          return
        }
        try {
          set({ hunt: markReady(hunt), screen: 'setup-ready' })
        } catch {
          set({ screen: 'setup-stage' })
        }
      },
      startAdventure: () => {
        const { hunt } = get()
        if (!hunt) {
          return
        }
        if (hunt.status === 'playing') {
          set({ screen: 'play', scanResult: null })
          return
        }
        set({
          hunt: startHunt(hunt),
          screen: 'play',
          scanResult: null,
        })
      },
      openScan: () => set({ screen: 'scan', scanResult: null }),
      closeScan: () => set({ screen: 'play', scanResult: null }),
      scanPayload: (payload) => {
        const { hunt, collection } = get()
        if (!hunt) {
          return null
        }
        const { hunt: nextHunt, result } = applyScan(hunt, payload)
        if (result.kind === 'cleared') {
          const awarded = awardTreasure(collection, nextHunt.id, new Date().toISOString())
          const alreadyInList = collection.some((item) => item.id === awarded.id)
          set({
            hunt: nextHunt,
            scanResult: result,
            screen: 'clear',
            lastAwardedId: awarded.id,
            collection: alreadyInList ? collection : [...collection, awarded],
          })
          return result
        }
        if (result.kind === 'advanced') {
          set({
            hunt: nextHunt,
            scanResult: result,
            screen: 'play',
          })
          return result
        }
        set({ scanResult: result })
        return result
      },
      clearScanResult: () => set({ scanResult: null }),
      openCollection: () => set({ screen: 'collection', scanResult: null }),
      finishClear: () => set({ screen: 'collection' }),
      abandonHunt: () =>
        set({
          hunt: null,
          screen: 'parent',
          setupIndex: 1,
          scanResult: null,
          lastAwardedId: null,
        }),
      setPracticeMode: (value) => set({ practiceMode: value }),
      reviewQr: () => {
        const { hunt } = get()
        if (!hunt) {
          return
        }
        set({ screen: 'setup-ready', setupIndex: 1 })
      },
    }),
    {
      name: 'cluegame-v1',
      partialize: (state) => ({
        hunt: state.hunt,
        collection: state.collection,
        screen: state.screen === 'scan' ? 'play' : state.screen,
        setupIndex: state.setupIndex,
        lastAwardedId: state.lastAwardedId,
        practiceMode: state.practiceMode,
      }),
    },
  ),
)

File: TakaraSagashi/TakaraSagashi/Store/GameStore.swift
import Foundation
import Observation

@Observable
@MainActor
final class GameStore {
    var screen: Screen = .home
    var hunt: Hunt?
    var setupIndex = 1
    var collection: [CollectedTreasure] = []
    var lastAwardedId: String?
    var scanResult: ScanResult?
    var practiceMode = false

    private let defaultsKey = "cluegame-native-v1"

    init() {
        restore()
        if screen == .scan {
            screen = .play
        }
    }

    func goTo(_ screen: Screen) {
        self.screen = screen
        scanResult = nil
        persist()
    }

    func beginSetup() {
        hunt = nil
        setupIndex = 1
        scanResult = nil
        screen = .setupCount
        persist()
    }

    func chooseStageCount(_ count: Int) {
        hunt = try? HuntEngine.createHunt(stageCount: count)
        setupIndex = 1
        screen = .setupStage
        persist()
    }

    func updateCurrentHint(_ hint: String) {
        guard let hunt else { return }
        self.hunt = try? HuntEngine.setStageHint(hunt, stageIndex: setupIndex, hint: hint)
        persist()
    }

    func nextSetupStage() {
        guard let hunt else { return }
        if setupIndex >= hunt.stageCount {
            finishSetup()
            return
        }
        setupIndex += 1
        persist()
    }

    func prevSetupStage() {
        if setupIndex <= 1 {
            screen = .setupCount
            persist()
            return
        }
        setupIndex -= 1
        persist()
    }

    func finishSetup() {
        guard let hunt else { return }
        do {
            self.hunt = try HuntEngine.markReady(hunt)
            screen = .setupReady
            persist()
        } catch {
            screen = .setupStage
        }
    }

    func startAdventure() {
        guard let hunt else { return }
        if hunt.status == .playing {
            screen = .play
            scanResult = nil
            persist()
            return
        }
        self.hunt = try? HuntEngine.startHunt(hunt)
        screen = .play
        scanResult = nil
        persist()
    }

    func openScan() {
        screen = .scan
        scanResult = nil
        persist()
    }

    func closeScan() {
        screen = .play
        scanResult = nil
        persist()
    }

    @discardableResult
    func scanPayload(_ payload: String) -> ScanResult? {
        guard let hunt else { return nil }
        let outcome = HuntEngine.applyScan(hunt, payload: payload)
        if outcome.result == .cleared {
            let awarded = TreasureCatalog.awardTreasure(collection: collection, huntId: outcome.hunt.id)
            if !collection.contains(where: { $0.id == awarded.id }) {
                collection.append(awarded)
            }
            self.hunt = outcome.hunt
            scanResult = outcome.result
            lastAwardedId = awarded.id
            screen = .clear
            persist()
            return outcome.result
        }
        if case .advanced = outcome.result {
            self.hunt = outcome.hunt
            scanResult = outcome.result
            screen = .play
            persist()
            return outcome.result
        }
        scanResult = outcome.result
        persist()
        return outcome.result
    }

    func clearScanResult() {
        scanResult = nil
    }

    func openCollection() {
        screen = .collection
        scanResult = nil
        persist()
    }

    func finishClear() {
        screen = .collection
        persist()
    }

    func abandonHunt() {
        hunt = nil
        setupIndex = 1
        scanResult = nil
        lastAwardedId = nil
        screen = .parent
        persist()
    }

    func reviewQR() {
        guard hunt != nil else { return }
        screen = .setupReady
        setupIndex = 1
        persist()
    }

    var currentStage: Stage? {
        hunt?.stages.first { $0.index == setupIndex }
    }

    var awardedTreasure: Treasure {
        if let lastAwardedId,
           let item = collection.first(where: { $0.id == lastAwardedId }) {
            return TreasureCatalog.treasure(id: item.treasureId)
        }
        return TreasureCatalog.all[0]
    }

    private struct Snapshot: Codable {
        var screen: Screen
        var hunt: Hunt?
        var setupIndex: Int
        var collection: [CollectedTreasure]
        var lastAwardedId: String?
        var practiceMode: Bool
    }

    private func persist() {
        let snapshot = Snapshot(
            screen: screen == .scan ? .play : screen,
            hunt: hunt,
            setupIndex: setupIndex,
            collection: collection,
            lastAwardedId: lastAwardedId,
            practiceMode: practiceMode
        )
        if let data = try? JSONEncoder().encode(snapshot) {
            UserDefaults.standard.set(data, forKey: defaultsKey)
        }
    }

    private func restore() {
        guard let data = UserDefaults.standard.data(forKey: defaultsKey),
              let snapshot = try? JSONDecoder().decode(Snapshot.self, from: data) else {
            return
        }
        screen = snapshot.screen
        hunt = snapshot.hunt
        setupIndex = snapshot.setupIndex
        collection = snapshot.collection
        lastAwardedId = snapshot.lastAwardedId
        practiceMode = snapshot.practiceMode
    }
}
